qrenderdoc/Windows/ShaderViewer.cpp

```diff
@@ -1797,15 +1809,17 @@ void ShaderViewer::applyForwardsChange()
           bool found = false;
           for(size_t i = 0; i < m_AccessedResources.size(); i++)
           {
-            if(c.after.GetBinding() == m_AccessedResources[i].GetBinding())
+            if(c.after.GetBinding() == m_AccessedResources[i].resource.GetBinding())
             {
               found = true;
+              if(m_AccessedResources[i].steps.indexOf(m_CurrentStateIdx - 1) < 0)
```

Comment:
I'll look into swapping that around, but I'm not certain that the scenario you describe is possible with DXBC - even a shader that performs a load immediately will spend the first instruction to set literals into a register for the load instruction. Do you know if the first instruction sample scenario is possible with DXIL or SPIR-V?

Reply:
Digging in a bit more, I'm not certain that the -1 is a problem. When running to a sample, the debugger steps to the instructions _after_ the sample occurs. So if step 0 were `ld_indexable`, running to sample to move to step 1. This makes sense, since the `SampleLoadGather` event occurs as a result of the instruction, not as it's about to occur. This is the same timeframe that the debugger provides accessed resources in a `ShaderDebugState` - an accessed resource will always occur on a step that has that event indicated. So even if a resource access occurred on the first instruction, the -1 _should_ be benign.

Note that there is an off-by-one error, somewhat intentionally, between the "run to" right click context menus for a resource and a step. Running to the previous/next access of a resource will follow the same rules as "run to sample" - moving to the step after that access occurred. "Go to instruction" does what it says on the cover, since those step numbers are provided in disassembly view. As an example, if you have a `ld_indexable` instruction on step 3, running to previous resource access would jump to step 4, while running "go to instruction" would jump to step 3. I'm not sure if that disparity is something you'd prefer reconciling (I can understand arguments either way), so I wanted to be sure to call it out here for discussion.

Reply:
I decided to eliminate the - 1 and have both resource accesses and steps align, so that running to them will yield the same results - going to the step immediately after the sample/load/gather instruction. I think ultimately trying to have then differ was going to be confusing for users.